src/array.c

```diff
@@ -629,156 +642,408 @@ static size_t limit_overallocation(jl_array_t *a, size_t alen, size_t newlen, si
-    size_t offset = a->offset;
-    offset += dec;
-    a->data = (char*)a->data + nb;
+    if (__unlikely(a->flags.isshared))
+        array_try_unshare(a);
+    jl_array_del_at_beg(a, 0, dec, n);
+}
+
+JL_DLLEXPORT void jl_array_del_end(jl_array_t *a, size_t dec)
+{
+    size_t n = jl_array_nrows(a);
+    if (__unlikely(n < dec))
+        jl_bounds_error_int((jl_value_t*)a, 0);
+    if (__unlikely(a->flags.isshared))
+        array_try_unshare(a);
+    jl_array_del_at_end(a, n - dec, dec, n);
+}
+
+JL_DLLEXPORT void jl_array_sizehint(jl_array_t *a, size_t sz)
+{
+    size_t n = jl_array_nrows(a);
+    if (sz <= n)
+        return;
+    size_t inc = sz - n;
+    jl_array_grow_end(a, inc);
+    a->nrows = n;
 #ifdef STORE_ARRAY_LEN
-    a->length -= dec;
+    a->length = n;
 #endif
-    a->nrows -= dec;
+}
 
-    // make sure offset doesn't grow forever due to deleting at beginning
-    // and growing at end
-    size_t newoffs = offset;
-    if (offset >= 13*a->maxsize/20) {
-        newoffs = 17*(a->maxsize - a->nrows)/100;
+JL_DLLEXPORT jl_array_t *jl_array_copy(jl_array_t *ary)
+{
+    size_t elsz = ary->elsize;
+    jl_array_t *new_ary = _new_array_(jl_typeof(ary), jl_array_ndims(ary),
+                                      &ary->nrows, !ary->flags.ptrarray, elsz);
+    memcpy(new_ary->data, ary->data, jl_array_len(ary) * elsz);
+    return new_ary;
+}
+
+// Copy element by element until we hit a young object, at which point
+// we can continue using `memmove`.
+static NOINLINE ssize_t jl_array_ptr_copy_forward(jl_value_t *owner,
+                                                  void **src_p, void **dest_p,
+                                                  ssize_t n)
+{
+    for (ssize_t i = 0; i < n; i++) {
+        void *val = src_p[i];
+        dest_p[i] = val;
+        // `val` is young or old-unmarked
+        if (val && !(jl_astaggedvalue(val)->bits.gc & GC_MARKED)) {
+            jl_gc_queue_root(owner);
+            return i;
+        }
     }
-#ifdef _P64
-    while (newoffs > (size_t)((uint32_t)-1)) {
-        newoffs = newoffs/2;
+    return n;
+}
+
+static NOINLINE ssize_t jl_array_ptr_copy_backward(jl_value_t *owner,
+                                                   void **src_p, void **dest_p,
+                                                   ssize_t n)
+{
+    for (ssize_t i = 0; i < n; i++) {
+        void *val = src_p[n - i - 1];
+        dest_p[n - i - 1] = val;
+        // `val` is young or old-unmarked
+        if (val && !(jl_astaggedvalue(val)->bits.gc & GC_MARKED)) {
+            jl_gc_queue_root(owner);
+            return i;
+        }
     }
-#endif
-    if (newoffs != offset) {
-        size_t anb = a->nrows*es;
-        size_t delta = (offset - newoffs)*es;
-        a->data = (char*)a->data - delta;
-        memmove(a->data, (char*)a->data + delta, anb);
+    return n;
+}
+
+// Unsafe, assume inbounds and that dest and src have the same eltype
+// `doffs` and `soffs` are zero based.
```

Comment:
I see. This like can be deleted, the function used to take offsets as arguments.